File: 프로그래머스/lv2/87946. 피로도/피로도.java
class Solution {
     
    int[][] dungeons;
    boolean[] visited;
    int answer = 0;
    
    public int solution(int k, int[][] DUNGEONS) {
        
        for (int i = 0; i < DUNGEONS.length; i++) {
            dungeons = DUNGEONS.clone();
        }
        visited = new boolean[dungeons.length];
        searchDungeons(k, 0);
        
        return answer;
    }
    
    /* DFS : 몇 개 던전 갈 수 있는지 탐색 */
    private void searchDungeons(int fatigue, int count) {        
        // 모든 던전을 다 탐색했을 때

        answer = Math.max(answer, count);

        for (int i = 0; i < dungeons.length; i++) { 
            if (visited[i]) continue;
            if (fatigue < dungeons[i][0]) continue;
            visited[i] = true;       
            searchDungeons(fatigue - dungeons[i][1], count + 1);
            visited[i] = false;
        }
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int solution(int k, vector<vector<int>> dungeons) { 
    
    int answer = 0;
    do {
        
        int curr = k; // 현재 피로도
        int count = 0; // 탐험한 던전 개수
        for (auto& dungeon : dungeons) {
            
            // 던전에 탐험이 가능한 경우
            if (curr >= dungeon[0]) {
                count++;
                curr -= dungeon[1];
            }
        }
        
        answer = max(answer, count);
        
    } while (next_permutation(dungeons.begin(), dungeons.end()));
    
    return answer;
}